code/headers/ssd1306.hpp

```diff
@@ -0,0 +1,120 @@
+        set_display_clock_div = 0xd5,
+        set_precharge = 0xd9,
+        set_multiplex = 0xa8,
+        set_low_column = 0x00,
+        set_high_column = 0x10,
+        set_start_line = 0x40,
+        memory_mode = 0x20,
+        column_addr = 0x21,
+        page_addr = 0x22,
+        com_scan_inc = 0xc0,
+        com_scan_dec = 0xc8,
+        seg_remap = 0xa0,
+        charge_pump = 0x8d,
+        external_vcc = 0x01,
+        switch_cap_vcc = 0x02,
+        activate_scroll = 0x2f,
+        deactivate_scroll = 0x2e,
+        set_vertical_scroll_area = 0xa3,
+        right_horizontal_scroll = 0x26,
+        left_horizontal_scroll = 0x27,
+        vertical_and_right_horizontal_scroll = 0x29,
+        vertical_and_left_horizontal_scroll = 0x2A
+    };
+
+    /// value to send over i2c before a command
+    constexpr uint8_t ssd1306_cmd_prefix = 0x80;
+
+    /// value to send over i2c before data
+    constexpr uint8_t ssd1306_data_prefix = 0x40;
+
+    /// SSD1306 chip initialization
+    constexpr const uint8_t ssd1306_initialization[] = {
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::display_off,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_display_clock_div,
+        0x80,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_multiplex,
+        0x3f,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_display_offset,
+        0x00,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_start_line | 0x00,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::charge_pump,
+        0x14,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::memory_mode,
+        0x00,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::seg_remap | 0x01,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::com_scan_dec,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_compins,
+        0x12,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_contrast,
+        0xcf,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_precharge,
+        0xf1,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::set_vcom_detect,
+        0x40,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::display_all_on_resume,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::normal_display,
+        ssd1306_cmd_prefix,
+        (uint8_t)ssd1306_commands::display_on};
+
+    class ssd1306_i2c_c {
+    protected:
+        /// The I2C bus
+        r2d2::i2c::i2c_bus_c bus;
+        /// The device address
+        uint8_t address;
+        /// The current cursor location in the controller
+        hwlib::xy cursor;
+
+    public:
+        // construct by providing the i2c channel
+        ssd1306_i2c_c(r2d2::i2c::i2c_bus_c &bus, const uint8_t &address);
+
+        /// send a command without data
+        void command(ssd1306_commands c);
```

Comment:
These function names are a bit inconsistent. Maybe the following is more logical?
 - `write_command`
- `write_data`
- `write_pixel`

Reply:
Not sure how they are inconsistent.

There is a `command` function for sending commands because they need to be prefixed
There is a `data` function for sending data because they need to be prefixed

The only weird function name is `pixels_byte_write` which, while weird, is logical considering it's writing a byte to a pixel. 

I think the names are consistent enough and clear in what they do. Please correct me if I'm wrong.

Reply:
`pixels_byte_write` doesn't describe to me what the function is doing, because I can't really imagine what "writing a byte to a pixel" really means. Maybe something like `data_at` or `data_to` would be logical. That's why I suggested the `write_` prefix, so you get `write_data_at` or `write_data_to`.

Reply:
Hmm, I get where you're coming from, but considering it accurately described what it does, and that this is a hardware class not to be used directly, I don't really see the issue.

I'm implementing hwlib::window on the actual instances, which means that the user only sees a `write(xy, color)` in their code, meaning that we get the best of both worlds where I get to be accurate about what the technical function does and the user doesn't need to know about the complexity behind it.

Reply:
Fair point. Let's leave it be.